darts/models/forecasting_model.py

```diff
@@ -101,6 +105,343 @@ def _build_forecast_series(self,
+
+        if not retrain and not self._fit_called:
+            fit_function(training_series.drop_after(start), target_series.drop_after(start), verbose=verbose)
+
+        for pred_time in iterator:
+            train = training_series.drop_after(pred_time)  # build the training series
+            target = target_series.drop_after(pred_time)  # build the target series
+            if (retrain):
+                fit_function(train, target)
+                pred = self.predict(forecast_horizon, **predict_kwargs)
+            else:
+                pred = self.predict(forecast_horizon, input_series=train, **predict_kwargs)
+            values.append(pred.values()[-1])  # store the N-th point
+            times.append(pred.end_time())  # store the N-th timestamp
+
+        forecast = TimeSeries.from_times_and_values(pd.DatetimeIndex(times), np.array(values))
+
+        return forecast
+
+    @classmethod
+    def gridsearch(model_class,
+                   parameters: dict,
+                   training_series: TimeSeries,
+                   target_series: Optional[TimeSeries] = None,
+                   forecast_horizon: Optional[int] = None,
+                   use_full_output_length: Optional[bool] = None,
+                   val_target_series: Optional[TimeSeries] = None,
+                   num_predictions: int = 10,
+                   use_fitted_values: bool = False,
+                   metric: Callable[[TimeSeries, TimeSeries], float] = metrics.mape,
+                   verbose=False):
+        """ A function for finding the best hyperparameters.
+
+        This function has 3 modes of operation: Expanding window mode, split mode and fitted value mode.
+        The three modes of operation evaluate every possible combination of hyperparameter values
+        provided in the `parameters` dictionary by instantiating the `model_class` subclass
+        of ForecastingModel with each combination, and returning the best-performing model with regards
+        to the `metric` function. The `metric` function is expected to return an error value,
+        thus the model resulting in the smallest `metric` output will be chosen.
+        The relationship of the training data and test data depends on the mode of operation.
+
+        Expanding window mode (activated when `forecast_horizon` is passed):
+        For every hyperparameter combination, the model is repeatedly trained and evaluated on different
+        splits of `training_series` and `target_series`. The number of splits is equal to `num_predictions`, and the
+        forecasting horizon used when making a prediction is `forecast_horizon`.
+        Note that the model is retrained for every single prediction, thus this mode is slower.
+
+        Split window mode (activated when `val_series` is passed):
+        This mode will be used when the `val_series` argument is passed.
+        For every hyperparameter combination, the model is trained on `training_series` + `target_series` and
```

Comment:
Just a thought: This naming of `training_series` and `target_series` is slightly confusing because usually in ML parlance the training set contains the targets as well. Could we maybe change to something better, like `train_feature_series` and `train_target_series`? Or even perhaps `X_train` and `y_train`. We would likely have to change in many places though.

Reply:
Yes, I think the naming here is tricky. Originally Guillaume used the names `covariate_series` and `target_series`, which I also thought could be misleading. How about `input_series` and `target_series`? None of the words are included in the set of {`train`, `validation`, `test`}, and, as supposed to using X and y, they preserve the notion of time series objects instead of any set of features and labels. I think X and y could also be somewhat confusing since we are not dealing with the conventional concept of labels in the sense that our 'labels' might also be included in the 'features' (which is also the case by default when training on univariate series). What do you think?

Reply:
I agree, I like your suggestion. Do you think we can do this change (in a consistent way, wherever needed across the codebase) in this PR, or do you prefer to wait for a next PR?

Reply:
Let's do it in a separate one!